src/networkmanager/cni.cpp

```diff
@@ -0,0 +1,951 @@
+              return OutputAccessConfigFromJson(aos::common::utils::CaseInsensitiveObjectWrapper(value));
+          });
+
+    Copy(outputAccess, firewall.mOutputAccess);
+}
+
+void ParseBandwidthConfig(const aos::common::utils::CaseInsensitiveObjectWrapper& plugin, BandwidthNetConf& bandwidth)
+{
+    bandwidth.mType         = plugin.GetValue<std::string>("type").c_str();
+    bandwidth.mIngressRate  = plugin.GetOptionalValue<uint64_t>("ingressRate").value_or(0);
+    bandwidth.mIngressBurst = plugin.GetOptionalValue<uint64_t>("ingressBurst").value_or(0);
+    bandwidth.mEgressRate   = plugin.GetOptionalValue<uint64_t>("egressRate").value_or(0);
+    bandwidth.mEgressBurst  = plugin.GetOptionalValue<uint64_t>("egressBurst").value_or(0);
+}
+
+} // namespace
+
+CNI::CNI(ExecItf* exec)
+    : mExec(exec)
+{
+}
+
+Error CNI::Init(const String& configDir)
+{
+    mConfigDir = configDir.CStr() + std::string("/results");
+
+    try {
+        if (!std::filesystem::create_directories(mConfigDir)) {
+            return ErrorEnum::eFailed;
+        }
+    } catch (const std::exception& e) {
+        return Error(ErrorEnum::eFailed, e.what());
+    }
+
+    return ErrorEnum::eNone;
+}
+
+RetWithError<Result> CNI::AddNetworkList(const NetworkConfigList& net, const RuntimeConf& rt)
+{
+    LOG_DBG() << "Add network list";
+
+    try {
+        auto prevResult = ResultToJSON(net.mPrevResult);
+        auto args       = ArgsAsString(rt, Action::eAdd);
+
+        prevResult = ExecuteBridgePlugin(net, prevResult, args);
+        prevResult = ExecuteDNSPlugin(net, rt, prevResult, args);
+        prevResult = ExecuteFirewallPlugin(net, prevResult, args);
+        prevResult = ExecuteBandwidthPlugin(net, prevResult, args);
+
+        auto result = ParsePrevResult(prevResult);
+
+        WriteCacheEntryToFile(
+            CreateCacheEntry(net, rt, prevResult), mConfigDir + "/" + net.mName.CStr() + "-" + rt.mContainerID.CStr());
+
+        return result;
+    } catch (const std::exception& e) {
+        return {Result {}, Error(ErrorEnum::eFailed, e.what())};
+    }
+}
+
+void CNI::WriteCacheEntryToFile(const std::string& cacheEntry, const std::string& cachePath) const
+{
+    std::ofstream cacheFile(cachePath);
+    if (!cacheFile.is_open()) {
+        throw std::runtime_error("Failed to open cache file");
+    }
+
+    cacheFile << cacheEntry;
+}
+
+Error CNI::DeleteNetworkList(const NetworkConfigList& net, const RuntimeConf& rt)
+{
+    LOG_DBG() << "Delete network list";
+
+    try {
+        auto prevResult = ResultToJSON(net.mPrevResult);
+        auto args       = ArgsAsString(rt, Action::eDel);
+
+        ExecuteBridgePlugin(net, prevResult, args);
+        ExecuteDNSPlugin(net, rt, prevResult, args);
+        ExecuteFirewallPlugin(net, prevResult, args);
+        ExecuteBandwidthPlugin(net, prevResult, args);
+
+        if (!std::filesystem::remove(mConfigDir + "/" + net.mName.CStr() + "-" + rt.mContainerID.CStr())) {
```

Comment:
Use std::filesystem::path to join path's

Reply:
Done